cpp/src/community/k_truss_impl.cuh

```diff
@@ -0,0 +1,951 @@
+  }
+};
+
+template <typename vertex_t, typename edge_t>
+struct extract_low_to_high_degree_edges_t {
+  __device__ thrust::optional<thrust::tuple<vertex_t, vertex_t>> operator()(vertex_t src,
+                                                                            vertex_t dst,
+                                                                            edge_t src_out_degree,
+                                                                            edge_t dst_out_degree,
+                                                                            thrust::nullopt_t) const
+  {
+    return (src_out_degree < dst_out_degree)
+             ? thrust::optional<thrust::tuple<vertex_t, vertex_t>>{thrust::make_tuple(src, dst)}
+             : (((src_out_degree == dst_out_degree) &&
+                 (src < dst) /* tie-breaking using vertex ID */)
+                  ? thrust::optional<thrust::tuple<vertex_t, vertex_t>>{thrust::make_tuple(src,
+                                                                                           dst)}
+                  : thrust::nullopt);
+  }
+};
+
+template <typename vertex_t, typename edge_t, bool generate_p_r>
+struct generate_p_r_and_q_r_from_p_q {
+  raft::device_span<size_t const> intersection_offsets{};
+  raft::device_span<vertex_t const> intersection_indices{};
+  raft::device_span<vertex_t const> invalid_srcs{};
+  raft::device_span<vertex_t const> invalid_dsts{};
+
+  __device__ thrust::tuple<vertex_t, vertex_t> operator()(edge_t i) const
+  {
+    auto itr = thrust::upper_bound(
+      thrust::seq, intersection_offsets.begin() + 1, intersection_offsets.end(), i);
+    auto idx = thrust::distance(intersection_offsets.begin() + 1, itr);
+
+    if constexpr (generate_p_r) {
+      return thrust::make_tuple(invalid_srcs[idx], intersection_indices[i]);
+
+    } else {
+      return thrust::make_tuple(invalid_dsts[idx], intersection_indices[i]);
+    }
+  }
+};
+}  // namespace
+
+template <typename vertex_t, typename edge_t, typename weight_t, bool multi_gpu>
+std::tuple<rmm::device_uvector<vertex_t>,
+           rmm::device_uvector<vertex_t>,
+           std::optional<rmm::device_uvector<weight_t>>>
+k_truss(raft::handle_t const& handle,
+        graph_view_t<vertex_t, edge_t, false, multi_gpu> const& graph_view,
+        std::optional<edge_property_view_t<edge_t, weight_t const*>> edge_weight_view,
+        edge_t k,
+        bool do_expensive_check)
+{
+  // 1. Check input arguments.
+
+  CUGRAPH_EXPECTS(!graph_view.has_edge_mask(), "unimplemented.");
+
+  CUGRAPH_EXPECTS(graph_view.is_symmetric(),
+                  "Invalid input arguments: K-truss currently supports undirected graphs only.");
+  CUGRAPH_EXPECTS(!graph_view.is_multigraph(),
+                  "Invalid input arguments: K-truss currently does not support multi-graphs.");
+
+  if (do_expensive_check) {
+    // nothing to do
+  }
+
+  // 2. Exclude self-loops (FIXME: better mask-out once we add masking support).
```

Comment:
Just FYI, see https://github.com/rapidsai/cugraph/pull/4261/files#diff-15b0881ee11d9efa224e7cf42b892ad9ba0b9788454f1757df97614083883097R182

So this FIXME is addressed in triangle counting. We may later refactor this code (may add something like `extract_dodg_graph` and call this utility function from multiple places rather than replicating).

Reply:
Got it. I am removing the FIXME